### ISSUE TITLE
With asCallback and finally, Unhandled Rejection is emitted.

### ISSUE THREAD

**Issue by NatalieWolfe** [petkaantonov/bluebird#1194]
1) What version of bluebird is the issue happening on?
Tested on 3.4.1, 3.3.0, and 2.10.2.

2) What platform and version? (For example Node.js 0.12 or Google Chrome 32)
Mac OS X, tested with Node 4.4.7 and 6.3.1.

3) Did this issue happen with earlier version of bluebird?
Yes, see answer to 1.

------------------
**promise.js**
```js
'use strict';

var Promise = require('bluebird');

process.on('unhandledRejection', function() {
  console.log('unhandledRejection');
})

Promise.reject('foo').asCallback(() => {
  console.log('asCallback');
}).finally(() => {
  console.log('finally');
});
```

**output**
```
$ node promise.js
asCallback
finally
unhandledRejection
```

Using the code snippet above, you'll see that `unhandledRejection` is emitted, despite the `asCallback` which will handle the rejection. I would expect the output to be just this:

**expected output**
```
$ node promise.js
asCallback
finally
```

If you remove the `.finally` bit so the promise sequence looks like this:
```js
Promise.reject('foo').asCallback(() => {
  console.log('asCallback');
});
```
Then the `unhandledRejection` event is no longer emitted:
```
$ node promise.js
asCallback
```
This seems correct to me, but is inconsistent with the `.finally` case.

**Comment by ben-page** [comment]
That's because `asCallback` is designed to make a function that can accept a callback **or return a Promise**. If you continue the Promise chain from `asCallback` (with `finally`, or `then`, etc), it thinks you want the Promise version.

Look at this example:

```javascript
function CallbackOrPromise(cb) {
    return Promise.reject(new Error('Reject!'))
        .asCallback(cb);
}

//callback version
CallbackOrPromise(function(err) {
    console.log('Error From Callback: ' + err.message);
});

//promise version
CallbackOrPromise()
    .catch(function(err) {
        console.log('Error From Promise: ' + err.message);
    });
```

What you are doing is like calling `CallbackOrPromise` without a catch:
```javascript
CallbackOrPromise()
    .then(function() {
        console.log('then'); //Unhandled rejection
    });
```

**Comment by NatalieWolfe** [comment]
I can understand the optimization of not creating a promise when `cb` is not a function, but when `cb` is a function which will handle the rejection, it seems like a bad idea. I agree that using a callback _and_ chaining off the promise is a strange use case, but it is one that I have now encountered in code that is not my own and that I can not change.

By having `asCallback` act differently than any other method it breaks expectations of how the promise chain should resolve. It means that without knowing the finer details of its execution, what looks like a normal chain of promises will resolve in unexpected ways.

The way `Promise.reject(...).asCallback(...).finally(...)` resolves is the same as this:
```js
var p = Promise.reject('foo');
p.then(() => console.log('resolved'), () => console.log('rejected'));
p.finally(() => {
  console.log('finally');
});
```
In the latter case it is explicitly obvious why that would cause an unhandled rejection, you have an explicitly branching promise and one of those branches does not handle the rejection. In the former `asCallback` case, the code is still branching, but as a hidden side effect. No other method does that.

**Comment by ben-page** [comment]
I think your example actually disproves your assertion that "`asCallback` acts differently". If you create multiple a Promise chain with multiple branches, each branch must handle the rejection. That's exactly how `asCallback` works. I see that may not be obvious at first, but that doesn't mean the design is bad.

It would be bad if Bluebird swallowed the Error instead. Imagine, someone used `then` instead of `finally`, but forgot to add `catch`. The `then` handler would never be called and the rejection would silently be ignored.

```javascript
CallbackOrPromise(function (err) {
    console.log('handled error');
})
.then(function () {
    console.log('then'); //never called, but thankfully the unhandled rejection would remind me that I need a catch
});
```

I think the current behavior is the correct one. Although, I don't actually see the point of supporting both callbacks and promises from the same call of `asCallback`. It may sense for Bluebird to throw an warning/error if a callback is passed to `asCallback` and the returned promise is continued. If it did, the author of the code you are referring to would have been warned about the misused of `asCallback`.

**Comment by NatalieWolfe** [comment]
I don't think we are going to agree on the merits of `asCallback`'s current design. :)

Adding the warning if both are used would be a good step in the right direction. It would at least make the intention of the function explicitly clear to anyone who uses it in an odd way. The error would be even better as it would completely prevent this bad use case.

**Comment by ben-page** [comment]
I'm not apart of the Bluebird team, but I'm fairly certain that you won't convince them that `asCallback` should swallow errors. However, I think we might have a case that `asCallback` should not support branching at all.

I'm propose that if `asCallback` is passed a function it doesn't return the Promise at all. So this:
```javascript
function callbackOrPromise(cb) {
    return Promise.resolve(true)
        .asCallback(cb);
}

//use as a function that returns a Promise
callbackOrPromise()
    .then(function () {
        console.log('from Promise');
    });

//use as a function that takes a callback
callbackOrPromise(function (err, value) {
    console.log('from Callback');
});

//can't have it both ways
callbackOrPromise(function (err) {
    console.log('from Callback');
})
.then(function () { //TypeError: Cannot read property 'then' of undefined
    console.log('then');
});
```

If someone really wanted to create branches, they still can do it explicitly:
```javascript
function callbackOrPromise(cb) {
    const p = Promise.resolve(true);
    p.asCallback(cb);
    return p;
}
```

**Comment by NatalieWolfe** [comment]
I am not advocating for `asCallback` to swallow errors. I am advocating for it to act the same as other promise methods (i.e. not branch) or be more explicit in how it differs.

**Comment by ben-page** [comment]
Ok, great. Do you agree with my proposal?

**Comment by NatalieWolfe** [comment]
Your proposal would resolve my issue, however I think not branching the chain would be an easier to adopt fix. Yours would outright break anyone who is using this odd case, while making `asCallback` chain instead of branch would bring it in line with the rest of the API.

**Comment by ben-page** [comment]
The documentation could be more clear on what `asCallback` does. In actuality, it doesn't create a new branch at all. In fact, doesn't even create a new promise in the chain. It just registers a callback to be run when the promise fulfills. That's it. It's enough to stop Bluebird from registering an unhandled rejection, but that's about it. Errors thrown in the callback are a regular uncaught exception (not a unhandled rejection).

So, what you are proposing is a significant change. As `asCallback` would need to create a new Promise that resolves to the return value of the callback and rejects if the callback throws. This of course, would also break anyone using this odd case, because the promise value might change.

**Comment by ben-page** [comment]
I made another PR with this behavior. When someone from Bluebird will weight in on this, I'll close one of them.
